DigitalLearningSolutions.Data/Services/LearningHubAuthApiUrlService.cs

```diff
@@ -0,0 +1,77 @@
+﻿namespace DigitalLearningSolutions.Data.Services
+{
+    using System.Web;
+    using DigitalLearningSolutions.Data.Helpers;
+    using Microsoft.Extensions.Configuration;
+
+    public interface ILearningHubAuthApiUrlService
+    {
+        string GetLoginUrlForDelegateAuthIdAndResourceUrl(string resourceUrl, int delegateAuthId);
+
+        string GetLinkingUrlForResource(int resourceReferenceId);
+    }
+
+    public class LearningHubAuthApiUrlService : ILearningHubAuthApiUrlService
+    {
+        private readonly IConfiguration config;
+        private readonly IGuidService guidService;
+        private readonly ILearningHubSsoSecurityService learningHubSsoSecurityService;
+
+        public LearningHubAuthApiUrlService(
+            IConfiguration config,
+            IGuidService guidService,
+            ILearningHubSsoSecurityService learningHubSsoSecurityService
+        )
+        {
+            this.config = config;
+            this.guidService = guidService;
+            this.learningHubSsoSecurityService = learningHubSsoSecurityService;
+        }
+
+        private string AuthBaseUrl => config.GetLearningHubAuthApiBaseUrl();
+        private string ClientCode => config.GetLearningHubAuthApiClientCode();
+
+        public string GetLoginUrlForDelegateAuthIdAndResourceUrl(string resourceUrl, int delegateAuthId)
+        {
+            var idHash = learningHubSsoSecurityService.GenerateHash(delegateAuthId.ToString());
+            var loginQueryString =
+                ComposeLoginQueryString(ClientCode, delegateAuthId, idHash, resourceUrl);
+
+            var loginEndpoint = config.GetLearningHubAuthApiLoginEndpoint();
+
+            return AuthBaseUrl + loginEndpoint + loginQueryString;
+        }
+
+        public string GetLinkingUrlForResource(int resourceReferenceId)
+        {
+            var state = ComposeCreateUserState(resourceReferenceId);
+            var stateHash = learningHubSsoSecurityService.GenerateHash(state);
+            var createUserQueryString = ComposeCreateUserQueryString(ClientCode, state, stateHash);
+
+            var linkingEndpoint = config.GetLearningHubAuthApiLinkingEndpoint();
+
+            return AuthBaseUrl + linkingEndpoint + createUserQueryString;
+        }
+
+        private string ComposeCreateUserState(int resourceReferenceId)
+        {
+            return guidService.NewGuid() + $"_refId:{resourceReferenceId}";

```

Comment:
This `GUID` should be stored in the users session also. See jira ticket for details.
Now you can inject the httpContext in here to do that. 
But you could also just pass a `Guid` into `GetLinkingUrlForResource` and do the `Guid` generation and "save to session" in the controller. This may simplify the setup and tests.